// @airtop/sdk/dist/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirtopClient = void 0;
const Client_1 = require("./api/resources/windows/client/Client");
const Client_2 = require("./api/resources/automations/client/Client");
const Client_3 = require("./api/resources/files/client/Client");
const Client_4 = require("./api/resources/profiles/client/Client");
const Client_5 = require("./api/resources/requests/client/Client");
const Client_6 = require("./api/resources/sessions/client/Client");
class AirtopClient {
    constructor(_options) {
        this._options = _options;
    }
    get windows() {
        var _a;
        return ((_a = this._windows) !== null && _a !== void 0 ? _a : (this._windows = new Client_1.Windows(this._options)));
    }
    get automations() {
        var _a;
        return ((_a = this._automations) !== null && _a !== void 0 ? _a : (this._automations = new Client_2.Automations(this._options)));
    }
    get files() {
        var _a;
        return ((_a = this._files) !== null && _a !== void 0 ? _a : (this._files = new Client_3.Files(this._options)));
    }
    get profiles() {
        var _a;
        return ((_a = this._profiles) !== null && _a !== void 0 ? _a : (this._profiles = new Client_4.Profiles(this._options)));
    }
    get requests() {
        var _a;
        return ((_a = this._requests) !== null && _a !== void 0 ? _a : (this._requests = new Client_5.Requests(this._options)));
    }
    get sessions() {
        var _a;
        return ((_a = this._sessions) !== null && _a !== void 0 ? _a : (this._sessions = new Client_6.Sessions(this._options)));
    }
}
exports.AirtopClient = AirtopClient;
